salt/minion.py

```diff
@@ -243,27 +243,45 @@ def resolve_dns(opts, fallback=True):
 
 
 def prep_ip_port(opts):
+    '''
+    parse host:port values from opts['master'] and return valid:
+        master: ip address or hostname as a string
+        master_port: (optional) master returner port as integer
+
+    e.g.:
+      - master: 'localhost:1234' -> {'master': 'localhost', 'master_port': 1234}
+      - master: '127.0.0.1:1234' -> {'master': '127.0.0.1', 'master_port' :1234}
+      - master: '[::1]:1234' -> {'master': '::1', 'master_port': 1234}
+      - master: 'fe80::a00:27ff:fedc:ba98' -> {'master': 'fe80::a00:27ff:fedc:ba98'}
+    '''
     ret = {}
     # Use given master IP if "ip_only" is set or if master_ip is an ipv6 address without
     # a port specified. The is_ipv6 check returns False if brackets are used in the IP
     # definition such as master: '[::1]:1234'.
     if opts['master_uri_format'] == 'ip_only' or salt.utils.network.is_ipv6(opts['master']):
-        ret['master'] = opts['master']
+        ret['master'] = str(ipaddress.ip_address(opts['master']))
     else:
-        ip_port = opts['master'].rsplit(':', 1)
-        if len(ip_port) == 1:
-            # e.g. master: mysaltmaster
-            ret['master'] = ip_port[0]
+        if opts['master'][0] == '[':
+            _elements = opts['master'].strip('[').split(']')
+            ret['master'] = str(ipaddress.IPv6Address(_elements[0]))
+            if len(_elements) == 2:
+                try:
+                    _port = _elements[1].split(':')[1]
+                except IndexError as exc:  # OK if there's no port specified, just pass
+                    log.debug("%s, unable to parse port number from opts['master'] IPv6 address: '%s'", exc, opts['master'])
+
         else:
-            # e.g. master: localhost:1234
-            # e.g. master: 127.0.0.1:1234
-            # e.g. master: [::1]:1234
-            # Strip off brackets for ipv6 support
-            ret['master'] = ip_port[0].strip('[]')
-
-            # Cast port back to an int! Otherwise a TypeError is thrown
-            # on some of the socket calls elsewhere in the minion and utils code.
-            ret['master_port'] = int(ip_port[1])
+            _elements = opts['master'].rsplit(':', 1)
+            try:
+                ret['master'] = str(ipaddress.ip_address(_elements[0]))
+            except ValueError as exc:
+                log.debug("%s: '%s' is not an IP address, treating as a hostname", exc, _elements[0])
+                ret['master'] = _elements[0]
+            try:
+                ret['master_port'] = int(_elements[1])
+            except IndexError as exc:  # OK if there's no port specified, just pass
+                log.debug("%s, unable to parse port number from opts['master'] IP address: '%s'", exc, opts['master'])
```

Comment:
And this is another example why we should do it in `ipaddress` class, rather then here. We would not have this kind of duplication.